File: flutter_app/android/app/src/dat/kotlin/com/personalos/amy_app/MetaGlassesBridge.kt
package com.personalos.amy_app

// Real Meta Wearables Device Access Toolkit (DAT) bridge — compiled only when
// the metaDat.enabled gradle property is true (see app/build.gradle.kts).
//
// Dart-facing contract (identical to the iOS bridge and the stub):
//   MethodChannel  amy/meta_glasses
//     isAvailable            -> {available: bool, reason?: str}
//     connect                -> {ok: bool, reason?: str, message?: str}
//     startStream {quality, frameRate} -> {ok, reason?}
//     stopStream             -> {ok}
//     triggerCapture         -> {ok, reason?}   (photo arrives as an event)
//     disconnect             -> {ok}
//     mockPair               -> {ok, reason?}   (Mock Device Kit, dev only)
//   EventChannel   amy/meta_glasses/events, map events:
//     {type: sessionState|streamState|registrationState, state: str-lowercase}
//     {type: capture, bytes: ByteArray(JPEG), takenAt: null, lat: null, lon: null}
//     {type: error, code: str, message: str}
//
// API surface verified against facebook/meta-wearables-dat-android v0.8.0
// (samples/CameraAccess + plugins/mwdat-android/skills). Two knowingly
// unverified spots are marked TODO(DAT-verify) below.

import android.app.Activity
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import java.io.ByteArrayOutputStream
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import com.meta.wearable.dat.camera.Stream
import com.meta.wearable.dat.camera.addStream
import com.meta.wearable.dat.camera.types.PhotoData
import com.meta.wearable.dat.camera.types.StreamConfiguration
import com.meta.wearable.dat.camera.types.VideoQuality
import com.meta.wearable.dat.core.Wearables
import com.meta.wearable.dat.core.selectors.AutoDeviceSelector
import com.meta.wearable.dat.core.session.DeviceSession
import com.meta.wearable.dat.core.types.Permission
import com.meta.wearable.dat.core.types.RegistrationState

class MetaGlassesBridge(private val activity: Activity) :
    MethodChannel.MethodCallHandler, EventChannel.StreamHandler {

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
    private var events: EventChannel.EventSink? = null
    private var methodChannel: MethodChannel? = null
    private var eventChannel: EventChannel? = null

    private var session: DeviceSession? = null
    private var stream: Stream? = null
    private var sessionStateJob: Job? = null
    private var streamStateJob: Job? = null
    private var streamErrorJob: Job? = null

    fun register(messenger: BinaryMessenger) {
        methodChannel = MethodChannel(messenger, "amy/meta_glasses").also {
            it.setMethodCallHandler(this)
        }
        eventChannel = EventChannel(messenger, "amy/meta_glasses/events").also {
            it.setStreamHandler(this)
        }
    }

    fun dispose() {
        teardown()
        scope.cancel()
        methodChannel?.setMethodCallHandler(null)
        eventChannel?.setStreamHandler(null)
        events = null
    }

    // --- EventChannel ------------------------------------------------------
    override fun onListen(arguments: Any?, sink: EventChannel.EventSink?) {
        events = sink
    }

    override fun onCancel(arguments: Any?) {
        events = null
    }

    private fun emit(event: Map<String, Any?>) {
        // Always deliver on the platform main thread (scope is Main-bound).
        scope.launch { events?.success(event) }
    }

    private fun emitError(code: String, message: String) =
        emit(mapOf("type" to "error", "code" to code, "message" to message))

    // --- MethodChannel -----------------------------------------------------
    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        try {
            when (call.method) {
                "isAvailable" -> result.success(mapOf("available" to true))
                "connect" -> connect(result)
                "startStream" -> startStream(
                    call.argument<String>("quality") ?: "medium",
                    call.argument<Int>("frameRate") ?: 15, result)
                "stopStream" -> { stopStreamInternal(); result.success(mapOf("ok" to true)) }
                "triggerCapture" -> triggerCapture(result)
                "disconnect" -> { teardown(); result.success(mapOf("ok" to true)) }
                "mockPair" -> mockPair(result)
                else -> result.notImplemented()
            }
        } catch (t: Throwable) {
            // A DAT exception must never crash the app: fail the call instead.
            result.success(mapOf("ok" to false, "reason" to "sdk_error",
                                 "message" to (t.message ?: t.toString())))
        }
    }

    // --- connect: registration -> permission -> session --------------------
    private fun connect(result: MethodChannel.Result) {
        scope.launch {
            try {
                if (Wearables.registrationState.value != RegistrationState.REGISTERED) {
                    // Deeplinks into the Meta AI app; the user approves there
                    // and this app shows up under glasses "App connections".
                    Wearables.startRegistration(activity)
                    result.success(mapOf(
                        "ok" to false, "reason" to "not_registered",
                        "message" to "Finish the approval in the Meta AI app, then tap Connect again."))
                    return@launch
                }

                // TODO(DAT-verify): the sample app only *checks* the wearables
                // camera permission; granting happens through the Meta AI
                // deeplink flow (startRegistration covers first-time grants).
                // If the SDK adds an explicit requestPermission API, call it
                // here on the denied path.
                var granted = false
                var checkFailed: String? = null
                Wearables.checkPermissionStatus(Permission.CAMERA)
                    .onSuccess { status ->
                        granted = status.toString().lowercase().contains("grant")
                    }
                    .onFailure { error, _ -> checkFailed = error.toString() }
                if (checkFailed != null) {
                    result.success(mapOf("ok" to false, "reason" to "sdk_error",
                                         "message" to "Permission check failed: $checkFailed"))
                    return@launch
                }
                if (!granted) {
                    result.success(mapOf(
                        "ok" to false, "reason" to "permission_denied",
                        "message" to "Grant camera access for this app in the Meta AI app (App connections)."))
                    return@launch
                }

                if (session == null) {
                    val s = Wearables.createSession(AutoDeviceSelector())
                    session = s
                    sessionStateJob = scope.launch {
                        s.state.collect { st ->
                            emit(mapOf("type" to "sessionState",
                                       "state" to st.toString().lowercase()))
                        }
                    }
                    s.start()
                }
                result.success(mapOf("ok" to true))
            } catch (t: Throwable) {
                emitError("connect_failed", t.message ?: t.toString())
                result.success(mapOf("ok" to false, "reason" to "sdk_error",
                                     "message" to (t.message ?: t.toString())))
            }
        }
    }

    // --- stream -------------------------------------------------------------
    private fun quality(name: String): VideoQuality = when (name.lowercase()) {
        "high" -> VideoQuality.HIGH
        "low" -> VideoQuality.LOW
        else -> VideoQuality.MEDIUM
    }

    private fun startStream(qualityName: String, frameRate: Int, result: MethodChannel.Result) {
        val s = session
        if (s == null) {
            result.success(mapOf("ok" to false, "reason" to "no_session",
                                 "message" to "Call connect first."))
            return
        }
        scope.launch {
            try {
                // Valid frame rates per DAT docs: 2, 7, 15, 24, 30.
                val fr = intArrayOf(2, 7, 15, 24, 30)
                    .minByOrNull { kotlin.math.abs(it - frameRate) } ?: 15
                s.addStream(StreamConfiguration(
                    videoQuality = quality(qualityName), frameRate = fr))
                    .onSuccess { added ->
                        stream = added
                        streamStateJob = scope.launch {
                            added.state.collect { st ->
                                emit(mapOf("type" to "streamState",
                                           "state" to st.toString().lowercase()))
                            }
                        }
                        streamErrorJob = scope.launch {
                            added.errorStream.collect { err ->
                                emitError("stream_error", err.toString())
                            }
                        }
                        added.start()
                        result.success(mapOf("ok" to true))
                    }
                    .onFailure { error, _ ->
                        result.success(mapOf("ok" to false, "reason" to "sdk_error",
                                             "message" to "addStream failed: $error"))
                    }
            } catch (t: Throwable) {
                result.success(mapOf("ok" to false, "reason" to "sdk_error",
                                     "message" to (t.message ?: t.toString())))
            }
        }
    }

    private fun stopStreamInternal() {
        streamStateJob?.cancel(); streamStateJob = null
        streamErrorJob?.cancel(); streamErrorJob = null
        try { stream?.stop() } catch (_: Throwable) {}
        stream = null
    }

    // --- capture ------------------------------------------------------------
    private fun triggerCapture(result: MethodChannel.Result) {
        val st = stream
        if (st == null) {
            result.success(mapOf("ok" to false, "reason" to "no_stream",
                                 "message" to "Start the stream first."))
            return
        }
        scope.launch {
            try {
                st.capturePhoto()
                    .onSuccess { photo -> handlePhoto(photo) }
                    .onFailure { error, _ -> emitError("capture_failed", error.toString()) }
                result.success(mapOf("ok" to true))
            } catch (t: Throwable) {
                result.success(mapOf("ok" to false, "reason" to "sdk_error",
                                     "message" to (t.message ?: t.toString())))
            }
        }
    }

    private fun handlePhoto(photo: PhotoData) {
        try {
            // DAT PhotoData carries no GPS and no wall-clock taken-at; per the
            // capture pipeline's rules we pass nulls through rather than
            // fabricating device-clock values.
            val jpeg: ByteArray? = when (photo) {
                is PhotoData.Bitmap -> {
                    val out = ByteArrayOutputStream()
                    photo.bitmap.compress(Bitmap.CompressFormat.JPEG, 90, out)
                    out.toByteArray()
                }
                // TODO(DAT-verify): HEIC payload field is `data` per the DAT
                // docs ("PhotoData exposes: data"); adjust here if 0.8.x names
                // it differently.
                is PhotoData.HEIC -> {
                    val bmp = BitmapFactory.decodeByteArray(photo.data, 0, photo.data.size)
                    if (bmp != null) {
                        val out = ByteArrayOutputStream()
                        bmp.compress(Bitmap.CompressFormat.JPEG, 90, out)
                        out.toByteArray()
                    } else photo.data // backend sniffs magic bytes; pass through
                }
            }
            if (jpeg == null || jpeg.isEmpty()) {
                emitError("capture_failed", "empty photo payload")
                return
            }
            emit(mapOf("type" to "capture", "bytes" to jpeg,
                       "takenAt" to null, "lat" to null, "lon" to null))
        } catch (t: Throwable) {
            emitError("capture_failed", t.message ?: t.toString())
        }
    }

    // --- mock device kit (dev/CI without physical glasses) ------------------
    private fun mockPair(result: MethodChannel.Result) {
        scope.launch {
            try {
                val kit = com.meta.wearable.dat.mockdevice.MockDeviceKit
                    .getInstance(activity.applicationContext)
                kit.enable()
                kit.pairGlasses(
                    com.meta.wearable.dat.mockdevice.api.GlassesModel.RAYBAN_META)
                result.success(mapOf("ok" to true))
            } catch (t: Throwable) {
                result.success(mapOf("ok" to false, "reason" to "sdk_error",
                                     "message" to (t.message ?: t.toString())))
            }
        }
    }

    // --- teardown ------------------------------------------------------------
    private fun teardown() {
        stopStreamInternal()
        sessionStateJob?.cancel(); sessionStateJob = null
        try { session?.stop() } catch (_: Throwable) {}
        // Stopped sessions are terminal in DAT: create a new one next connect.
        session = null
    }
}

File: flutter_app/ios/Runner/MetaGlassesBridge.swift
import Flutter
import Foundation

// Real Meta Wearables Device Access Toolkit (DAT) bridge for iOS.
//
// The DAT SDK is added via Swift Package Manager (Xcode → Add Package →
// https://github.com/facebook/meta-wearables-dat-ios → MWDATCore +
// MWDATCamera [+ MWDATMockDevice for testing]). Until that package is added,
// the canImport guards below compile this file into a stub that answers
// every call with reason=sdk_not_bundled — so plain `flutter build ios`
// keeps working unchanged. See flutter_app/META_GLASSES.md.
//
// Dart-facing contract (identical to the Android bridge):
//   MethodChannel  amy/meta_glasses
//     isAvailable / connect / startStream{quality,frameRate} / stopStream /
//     triggerCapture / disconnect / mockPair
//   EventChannel   amy/meta_glasses/events:
//     {type: sessionState|streamState|registrationState, state: <lowercase>}
//     {type: capture, bytes: FlutterStandardTypedData(JPEG),
//      takenAt: nil, lat: nil, lon: nil}
//     {type: error, code, message}
//
// API surface verified against facebook/meta-wearables-dat-ios
// (samples/CameraAccess + plugins/mwdat-ios/skills). Knowingly unverified
// spots are marked TODO(DAT-verify).

#if canImport(MWDATCore)
import MWDATCore
#endif
#if canImport(MWDATCamera)
import MWDATCamera
#endif
#if canImport(MWDATMockDevice)
import MWDATMockDevice
#endif

final class MetaGlassesBridge: NSObject, FlutterStreamHandler {
  private var eventSink: FlutterEventSink?

  static func register(messenger: FlutterBinaryMessenger) {
    let bridge = MetaGlassesBridge()
    let method = FlutterMethodChannel(name: "amy/meta_glasses", binaryMessenger: messenger)
    let events = FlutterEventChannel(name: "amy/meta_glasses/events", binaryMessenger: messenger)
    events.setStreamHandler(bridge)
    method.setMethodCallHandler { call, result in
      bridge.handle(call, result: result)
    }
    // Keep the bridge alive for the app's lifetime.
    objc_setAssociatedObject(method, "amy_glasses_bridge", bridge, .OBJC_ASSOCIATION_RETAIN)
  }

  // --- FlutterStreamHandler ------------------------------------------------
  func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    eventSink = events
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    eventSink = nil
    return nil
  }

  private func emit(_ event: [String: Any?]) {
    DispatchQueue.main.async { [weak self] in
      self?.eventSink?(event as [String: Any?])
    }
  }

  private func emitError(_ code: String, _ message: String) {
    emit(["type": "error", "code": code, "message": message])
  }

  private func ok() -> [String: Any?] { ["ok": true] }

  private func fail(_ reason: String, _ message: String) -> [String: Any?] {
    ["ok": false, "reason": reason, "message": message]
  }

#if canImport(MWDATCore) && canImport(MWDATCamera)
  // ==========================================================================
  // REAL IMPLEMENTATION (SPM package present)
  // ==========================================================================
  private static var configured = false

  // TODO(DAT-verify): session type is `DeviceSession` per DAT docs; adjust if
  // the SPM module names it differently.
  private var session: DeviceSession?
  private var stream: MWDATCamera.Stream?
  private var sessionStateTask: Task<Void, Never>?
  private var stateListenerToken: AnyListenerToken?
  private var errorListenerToken: AnyListenerToken?
  private var photoDataListenerToken: AnyListenerToken?

  private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "isAvailable":
      result(["available": true])
    case "connect":
      connect(result)
    case "startStream":
      let args = call.arguments as? [String: Any] ?? [:]
      startStream(quality: args["quality"] as? String ?? "medium",
                  frameRate: args["frameRate"] as? Int ?? 15, result: result)
    case "stopStream":
      stopStreamInternal()
      result(ok())
    case "triggerCapture":
      triggerCapture(result)
    case "disconnect":
      teardown()
      result(ok())
    case "mockPair":
      mockPair(result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func connect(_ result: @escaping FlutterResult) {
    Task { [weak self] in
      guard let self else { return }
      do {
        if !Self.configured {
          Wearables.configure()   // must run before any other DAT call
          Self.configured = true
        }
        let wearables = Wearables.shared

        // Registration gate: first value of the registration state stream.
        var registered = false
        for await state in wearables.registrationStateStream() {
          registered = (state == .registered)
          self.emit(["type": "registrationState",
                     "state": String(describing: state).lowercased()])
          break
        }
        if !registered {
          // Deeplinks to the Meta AI app; user approves there and this app
          // appears under the glasses' "App connections".
          try? await wearables.startRegistration()
          result(self.fail("not_registered",
                           "Finish the approval in the Meta AI app, then tap Connect again."))
          return
        }

        // NOTE: unlike Android, the iOS sample exposes no explicit wearables
        // camera-permission check API — permission problems surface as the
        // session/stream pausing or erroring, which we forward as events.
        if self.session == nil {
          let selector = AutoDeviceSelector(wearables: wearables)
          let s = try wearables.createSession(deviceSelector: selector)
          self.session = s
          self.sessionStateTask = Task { [weak self] in
            for await st in s.stateStream() {
              self?.emit(["type": "sessionState",
                          "state": String(describing: st).lowercased()])
            }
          }
          try s.start()
        }
        result(self.ok())
      } catch {
        self.emitError("connect_failed", error.localizedDescription)
        result(self.fail("sdk_error", error.localizedDescription))
      }
    }
  }

  private func resolution(_ name: String) -> StreamingResolution {
    switch name.lowercased() {
    case "high": return .high
    case "low": return .low
    default: return .medium
    }
  }

  private func startStream(quality: String, frameRate: Int, result: @escaping FlutterResult) {
    guard let session else {
      result(fail("no_session", "Call connect first."))
      return
    }
    do {
      // Valid frame rates per DAT docs: 2, 7, 15, 24, 30.
      let valid = [2, 7, 15, 24, 30]
      let fr = valid.min(by: { abs($0 - frameRate) < abs($1 - frameRate) }) ?? 15
      let config = StreamConfiguration(videoCodec: .raw,
                                       resolution: resolution(quality),
                                       frameRate: fr)
      guard let newStream = try session.addStream(config: config) else {
        result(fail("sdk_error", "addStream returned nil"))
        return
      }
      stream = newStream
      stateListenerToken = newStream.statePublisher.listen { [weak self] state in
        self?.emit(["type": "streamState",
                    "state": String(describing: state).lowercased()])
      }
      photoDataListenerToken = newStream.photoDataPublisher.listen { [weak self] photoData in
        // DAT photos carry no GPS / wall-clock taken-at: pass nils through,
        // never fabricate device-clock values (same rule as Android).
        self?.emit(["type": "capture",
                    "bytes": FlutterStandardTypedData(bytes: photoData.data),
                    "takenAt": nil, "lat": nil, "lon": nil])
      }
      newStream.start()
      result(ok())
    } catch {
      result(fail("sdk_error", error.localizedDescription))
    }
  }

  private func triggerCapture(_ result: @escaping FlutterResult) {
    guard let stream else {
      result(fail("no_stream", "Start the stream first."))
      return
    }
    let accepted = stream.capturePhoto(format: .jpeg)
    if accepted {
      result(ok())   // photo arrives via photoDataPublisher -> capture event
    } else {
      result(fail("capture_failed", "capturePhoto was not accepted by the stream"))
    }
  }

  private func stopStreamInternal() {
    stateListenerToken = nil
    errorListenerToken = nil
    photoDataListenerToken = nil
    stream?.stop()
    stream = nil
  }

  private func teardown() {
    stopStreamInternal()
    sessionStateTask?.cancel()
    sessionStateTask = nil
    session?.stop()
    // Stopped sessions are terminal in DAT: create a new one next connect.
    session = nil
  }

  private func mockPair(_ result: @escaping FlutterResult) {
#if canImport(MWDATMockDevice)
    // TODO(DAT-verify): the sample injects a MockDeviceKitInterface; direct
    // construction assumed here — adjust to the package's factory if needed.
    let kit = MockDeviceKit()
    kit.enable()
    _ = kit.pairGlasses(model: .rayBanMeta)
    result(ok())
#else
    result(fail("sdk_not_bundled", "Add the MWDATMockDevice product to Runner to use the mock device kit."))
#endif
  }

#else
  // ==========================================================================
  // STUB (SPM package not added) — keeps plain builds green; Dart sees the
  // feature as unavailable with reason=sdk_not_bundled.
  // ==========================================================================
  private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "isAvailable":
      result(["available": false, "reason": "sdk_not_bundled"])
    case "connect", "startStream", "stopStream", "triggerCapture",
         "disconnect", "mockPair":
      result(fail("sdk_not_bundled",
                  "This build was made without the Meta DAT SDK (add the SPM package in Xcode)."))
    default:
      result(FlutterMethodNotImplemented)
    }
  }
#endif
}
